File: companies/Walmart Labs/116.populating-next-right-pointers-in-each-node.java
/*
 * LeetCode provides:
 * class Node {
 *     public int val;
 *     public Node left;
 *     public Node right;
 *     public Node next;
 * }
 */

/**
 * Algorithm:
 * In a perfect binary tree, every internal node has both children. Walk each
 * level using already-created next pointers, linking node.left -> node.right
 * and node.right -> node.next.left when a neighbor exists.
 *
 * Java data structures:
 * No queue is needed; next pointers act as the level traversal structure.
 *
 * Complexity:
 * Time O(n), extra space O(1).
 */
class Solution {
    public Node connect(Node root) {
        Node leftmost = root;
        while (leftmost != null && leftmost.left != null) {
            Node node = leftmost;
            while (node != null) {
                node.left.next = node.right;
                if (node.next != null) {
                    node.right.next = node.next.left;
                }
                node = node.next;
            }
            leftmost = leftmost.left;
        }
        return root;
    }
}

File: companies/Uber/116.populating-next-right-pointers-in-each-node.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;

// Definition for a Node.
// class Node {
// public:
//     int val;
//     Node* left;
//     Node* right;
//     Node* next;
//     Node() : val(0), left(nullptr), right(nullptr), next(nullptr) {}
//     Node(int _val) : val(_val), left(nullptr), right(nullptr), next(nullptr) {}
//     Node(int _val, Node* _left, Node* _right, Node* _next)
//         : val(_val), left(_left), right(_right), next(_next) {}
// };


class Solution {
public:
    Node* connect(Node* root) {
        /*
        Approach:
        The tree is perfect, so every internal node has both children. Walk each
        level using existing next pointers. Link node->left to node->right, and
        node->right to node->next->left when a neighbor exists.

        C++ notes:
        Node* next pointers are mutated in place; no queue is needed.

        Complexity: O(n) time and O(1) extra space.
        */
        Node* leftmost = root;
        while (leftmost && leftmost->left) {
            Node* node = leftmost;
            while (node) {
                node->left->next = node->right;
                if (node->next) node->right->next = node->next->left;
                node = node->next;
            }
            leftmost = leftmost->left;
        }
        return root;
    }
};
